### DISCUSSION TITLE
Emoji Picker ⚒️

### DISCUSSION THREAD

**Discussion by zzeleznick** [johnlindquist/kit#107] [Share]
## Emoji Picker ⚒️

While you can press `ctrl + cmd + space` to launch the native emoji keyboard, or select an emoji from newer Mac's touch-bar, sometimes it's fun to build your own!

Here, you can find two demos – the first uses the panel html feature to enable regex filtering of the emojis, whereas the second uses the built in filtering for choices.

While currently the UI's don't match, perhaps in future releases, the same look could be achieved 😄 

### Panel Demo

https://user-images.githubusercontent.com/5779832/113371307-e4ae6800-931a-11eb-8e31-bfb8d93aec5d.mp4

### Choices Demo

https://user-images.githubusercontent.com/5779832/113371326-e7a95880-931a-11eb-87d6-5f93535ec9a6.mp4

### Code
<details>
<summary>Click for code</summary>

```js
// Menu: Kit Emojis
// Description: View + Copy Emojis
// Author: Zach Zeleznick
// Twitter: @zzxiv
// Shortcut: cmd e

const emojisDB = db("emojis", { emojis: {} });
const emojisRef = emojisDB.get("emojis");

// NOTE: Should extract this into a lib since emojis db used in kit-discussions ...
const fetchEmojis = async () => {
  // Could install and use as an npm package, but we just need a k-v map ...
  const emojiURL = 'https://raw.githubusercontent.com/omnidan/node-emoji/master/lib/emoji.json';
  const response = await get(emojiURL);
  const emojis = response.data;
  emojisDB.set("emojis", emojis).write();
}

const setupEmojis = async () => {
  const emojis = emojisRef.value();
  if (!emojis || !Object.keys(emojis).length) {
    await fetchEmojis()
  }
  return emojis
}

const createRegEx = (input = '') => {
  input = input.trim().toLowerCase()
  // NOTE: don't check length here for snappy ux
  // input = input.length < 3 ? '' : input
  let matcher = input
  try {
    matcher = new RegExp(input)
  } catch (err) {
    console.warn("Invalid expression", input)
  }
  return matcher
}

const showEmojis = (emojis) => (input) => {
  const matcher = createRegEx(input)
  const inner = Object.entries(emojis)
    .filter(([k,v]) => k.match(matcher) !== null)
    .map(([k,v]) => `<div class="flex h-10 w-full justify-start items-center">
        <div class="text-base font-bold font-sans mr-8"> ${v} </div>
        <div class="text-xs font-mono"> ${k} </div>
    </div>`).join('\n');
  const html = `
  <div class="grid grid-cols-1">
   ${inner}
  </div>`
  return html
}

const buildEmojisRxPanel = async () => {
  const emojis = await setupEmojis();
  await arg({
    message: "Search for emoji:",
    input: "",
  }, showEmojis(emojis));
}

const buildEmojisChoices = async () => {
  const emojis = await setupEmojis();
  const choices = Object.entries(emojis)
    .map(([k,v]) => {
    return {
      name: k,
      value: v,
      html: `<div> ${v} </div>`
    }
  });
  const emoji = await arg("Search for emoji:", choices);
  copy(emoji);
}

const panel = true // would be nice to set based on whether shift is pressed with shortcut cmd

panel ? await buildEmojisRxPanel() : await buildEmojisChoices();


// NOTE: tabs don't play nicely with choices + panels

// onTab("Choices", buildChoicesEmojis);
// onTab("Panel", buildReactivePrompt);

```
</details>

### Appendix

I've been a fan of emojipedia.org, but now that I learned the mac keyboard shortcut (`ctrl + cmd + space`), I might use it more often...

<table>
<tr>
<td>
<img width="476" alt="Emoji Keyboard" src="https://user-images.githubusercontent.com/5779832/113372024-897d7500-931c-11eb-90be-da508ed18d28.png">
</td>
<td>
<img width="455" alt="Emoji Keyboard Search" src="https://user-images.githubusercontent.com/5779832/113372021-86828480-931c-11eb-8e4f-d714ef39c46e.png">
</td>
</tr>
<tr>
<td>
 Mac Emoji Keyboard on terminal cursor
</td>
<td>
 Mac Emoji Keyboard supports search
</td>
</tr>
</table>

<table>
<tr>
<td>
<img width="621" alt="Emoji Character Viewer" src="https://user-images.githubusercontent.com/5779832/113372177-e2e5a400-931c-11eb-821e-97992adb56e9.png">
</td>
<td>
<img width="616" alt="Emoji Viewer Search" src="https://user-images.githubusercontent.com/5779832/113372171-dfeab380-931c-11eb-8976-7f84725b8b48.png">
</td>
</tr>
<tr>
<td>
 Mac Emoji Character Viewer Menu
</td>
<td>
 Mac Character Viewer Menu also supports search
</td>
</tr>
</table>

**PS.** For future work, it could be cool to extend the emoji list (see instructions on [node-emoji](https://www.npmjs.com/package/node-emoji)) to support adding custom emojis.

**Comment by johnlindquist** [comment]
I'll look in to the tabs/choices/panel issue.

At the end of the script, you can:

let {setSelectedText} = await kit('text')

setSelectedText(emoji)

to paste
